write react-toastify study (#3)
jackbsteinberg/std-toast: study-group/Toast-Implementation-Studies.md

```kotlin
val text = "Hello toast!"
val duration = Toast.LENGTH_SHORT

val toast = Toast.makeText(applicationContext, text, duration)
toast.show()
```

This creates a toast object filled with a context, some text, and a duration, which is triggered to display by the `toast.show()` method.
This will show up on the bottom of the screen by default.
To control the position of the toast, set the gravity, like this:

```kotlin
toast.setGravity(Gravity.TOP or Gravity.LEFT, 0, 0)
```

This sample will move the toast to the top left of the screen.
The final two parameters are used to offset the toast from its center of gravity, so this:

```kotlin
toast.setGravity(Gravity.TOP or Gravity.LEFT, 10, 10)
```

Will show the toast further down and further right.
The toast can also be injected with a custom layout, which requires more complex Kotlin, explained [here](https://developer.android.com/guide/topics/ui/notifiers/toasts#CustomToastView).
The `text`, `view`, `gravity`, and `duration` of the toast are all the elements the developer can control when working with this very straightforward component.

### Notable Features & Details

#### Gravity

> The library takes an interesting approach to positioning, by defining letting the user define the "center of gravity" of the element, along with offset values for the axes of gravity.

Reviewer: ```suggestion
The library takes an interesting approach to positioning, by letting the user define the "center of gravity" of the element, along with offset values for the axes of gravity.
```

> The library takes an interesting approach to positioning, by defining letting the user define the "center of gravity" of the element, along with offset values for the axes of gravity.

Reviewer: Missing semantic line breaks. I think I see the problem. You can and should break within sentences, e.g. this should probably be 3 lines.

Author: I'm still getting used to this semantic line break pattern, especially when in the middle of sentences.
I'll add this in now.

> The library takes an interesting approach to positioning, by defining letting the user define the "center of gravity" of the element, along with offset values for the axes of gravity.

Reviewer: Gravity (not center of gravity) is a standard android UI concept, maybe just point that out.

#### Duration Constants
The library comes with baked in constants for displaying the toast for long and short periods of time.

### Takeaways
- High opinion, low control.
- The API for the `Toast` object provides getters and setters for each of `text`, `view`, `gravity`, `offset` and `duration`.
- Beyond getters and setters, the only control the object allows is `.makeText()`, `.show()` and `.cancel()`.

## [react-toastify](https://github.com/fkhadra/react-toastify)
`react-toastify` is a React component library by author [@fkhadra](https://github.com/fkhadra) for rich, extensible notification toasts.

### Design & Sample Code

To use a `react-toastify` toast requires some simple JavaScript:

```js
class App extends Component {
  notify = () => toast("Very simple toast!");

  render(){
    return (
      <div>
        <button onClick={this.notify}>Notify !</button>
        <ToastContainer />
      </div>
    );
  }
}
```

`react-toastify` uses the explicit component pattern, by having the developer render a React `ToastContainer` component (the recommendation is to put this in the application root).

> This single component will render all toasts called from JavaScript, and can be configured with global options.

Reviewer: don't think you need a "," here.

Author: ```suggestion
This single component will render all toasts called from JavaScript and can be configured with global options.
```

However, the library allows the developer a lot of freedom, supporting multiple `ToastContainer`s as well as no `ToastContainer`, though one is recommended.
When the `toast()` is called with a message and a props object, it is displayed in the toast container, with the toast props overriding the `ToastContainer` props, like so:

```js
class App extends Component {
  tenSecondToast = () => toast("This toast will self destruct in 10...", { autoClose: 10000 });
  normalToast = () => toast("This is a normal toast");

  render(){
    return (
      <div>

>         <button onClick={this.tenSecondToast}>Close after 10 seconds</button>

Reviewer: ```suggestion
        <button onClick={this.tenSecondToast}>Close after 10000ms seconds</button>
```
I think it's easier on the reader to use the same numbers as the code with the API's units.

Author: You're right that makes sense

>         <button onClick={this.normalToast}>Close after default amount of time</button>

Reviewer: This example was unclear to me until I read the text below. I couldn't tell if "default" was referring to 5000 or a more global default.

I would suggest
```suggestion
        <button onClick={this.normalToast}>Close after 5000ms</button>
```

Author: ```suggestion
        <button onClick={this.normalToast}>Close after 5000ms (default)</button>
```

        <ToastContainer autoClose={5000} />
      </div>
    );
  }
}
```

> The first button creates a toast that dismisses in 10 seconds, overriding the `ToastContainer` default.

Reviewer: I'm not sure that this text is adding much value as the example also explains the outcome. Perhaps move anything extra here into the example?

The second provides no `autoClose` value, so it dismisses in the default 5 seconds.

> For some of the more common controls, the toast can be configured like so:

Reviewer: new paragraph?

    onClose: ({ props }) => window.alert('I warned you') // props: all props passed to rendered toast component
  });

  render() {
    return (
      <button onClick={this.topRightWarning}>Warn me</button>
      <ToastContainer autoClose={5000} />
    );
  }
}
```

`react-toastify` supports built-in themes for toasts, such as `success`, `info`, and `warning`, among others.
It also provides intuitive positioning, that can be set from the toast or configured on the `ToastContainer`.
Additionally, callbacks like `onClose` and `onOpen` allow the toast to have intelligent effects when the user interacts with it.

### Notable Features & Details

#### Toast ID
the `toast()` function returns an ID of the toast it creates, which can be used to act upon that toast.
For instance, with the toast ID the toast can be dismissed, updated, and checked to see if it's active.

#### Mobile
On mobile toasts will automatically take all available width, and all toasts are swipeable / draggable to dismiss.

#### Styling
Library provides fields `className`, `toastClassName`, and, `progressClassName` for the developer to pass in CSS classes.

#### Render Custom Component

> To customize the view the `toast` function can take in a React component, which will receive the function required to close the toast as props (though it won't get the toast ID).

Reviewer: Can you rephrase this? I don't know what this means.

Author: This was unclear, rephrasing now

#### cssTransitionHelper
The library provides a `cssTransitionHelper` helper function to customize animations without having to work directly with `react-transition-group`.
The helper allows the dev to set classes and durations for `enter` and `exit`, as well as make transitions position specific.

### Takeaways
- Extremely unopinionated and configurable; the developer has near total control.
- Component pattern is optional, only needs one component, but allows multiple.
- Close button is default, but can be replaced with a custom close button.
  - Separate from passing in a custom React Component.
- The toast can be configured to display after a certain delay.
  - I'm not sure what the advantage this has over firing the toast in a setTimeout.

> - Passing null into a property resets it to default (useful for `update()`).

Reviewer: What is update()? It's not referenced above.

Author: Was not clear enough with this, changing it now

This will show up on the bottom of the screen by default.
To control the position of the toast, set the gravity, like this:

```kotlin
toast.setGravity(Gravity.TOP or Gravity.LEFT, 0, 0)
```

This sample will move the toast to the top left of the screen.
The final two parameters are used to offset the toast from its center of gravity, so this:

```kotlin
toast.setGravity(Gravity.TOP or Gravity.LEFT, 10, 10)
```

Will show the toast further down and further right.
The toast can also be injected with a custom layout, which requires more complex Kotlin, explained [here](https://developer.android.com/guide/topics/ui/notifiers/toasts#CustomToastView).
The `text`, `view`, `gravity`, and `duration` of the toast are all the elements the developer can control when working with this very straightforward component.

### Notable Features & Details

#### Gravity
The library takes the Android approach to positioning,
by letting the user define the Gravity of the element,
along with offset values for the axes.

#### Duration Constants
The library comes with baked in constants for displaying the toast for long and short periods of time.

### Takeaways
- High opinion, low control.

> - The API for the `Toast` object provides getters and setters for each of `text`, `view`, `gravity`, `offset` and `duration`.

Reviewer 2: Does this mean they can be changed dynamically, while the toast is displaying? I hadn't thought of that before, but I guess it makes sense. I'm not sure if it's worth mentioning explicitly, but I found it interesting.

Author: To be honest, I'm not sure. The way the toast itself works is that you create and configure the object, then show it. 

What I think is probably more likely is that `show()` just shows the toast with the previously configured options, and setting the options mid-display will only affect the toast the next time you call `show()`.

> - Beyond getters and setters, the only control the object allows is `.makeText()`, `.show()` and `.cancel()`.

Reviewer 2: I'd omit `makeText()` since it's the factory method, not really about controlling.

Author: Makes sense, on it
```suggestion
- Beyond getters and setters, the only controls the object allows are `.show()` and `.cancel()`.
```

## [react-toastify](https://github.com/fkhadra/react-toastify)
`react-toastify` is a React component library by author [@fkhadra](https://github.com/fkhadra) for rich, extensible notification toasts.

### Design & Sample Code

> To use a `react-toastify` toast requires some simple JavaScript:

Reviewer 2: Avoid using "simple"; it's a judgement on the reader's ability that can feel bad if they find the code harder to understand.

Author: Completely understand, definitely doesn't need to be there
```suggestion
Using a `react-toastify` toast requires some JavaScript:
```

> To use a `react-toastify` toast requires some simple JavaScript:

Reviewer 2: Maybe "react code" instead of "JavaScript" since this uses JSX.

> ```js

Reviewer 2: I'm curious if using `jsx` instead of `js` give different results for syntax highlighting.

Author: Ooh it does, that's good to know

class App extends Component {
  notify = () => toast("Very simple toast!");

  render(){
    return (
      <div>
        <button onClick={this.notify}>Notify !</button>

>         <ToastContainer />

Reviewer 2: Maybe show the importing of toast/ToastContainer?

Author: Good call

      </div>
    );
  }
}
```

`react-toastify` uses the explicit component pattern, by having the developer render a React `ToastContainer` component (the recommendation is to put this in the application root).
This single component will render all toasts called from JavaScript and can be configured with global options.
However, the library allows the developer a lot of freedom, supporting multiple `ToastContainer`s as well as no `ToastContainer`, though one is recommended.

When the `toast()` is called with a message and a props object, it is displayed in the toast container, with the toast props overriding the `ToastContainer` props, like so:

```js
class App extends Component {
  // Creating a toast that will close in 10000ms.
  tenSecondToast = () => toast("This toast will self destruct in 10...", { autoClose: 10000 });
  // Creating a toast that will close in 5000ms (the default).
  normalToast = () => toast("This is a normal toast");

  render(){
    return (
      <div>

>         <button onClick={this.tenSecondToast}>Close after 10000ms seconds</button>

Reviewer 2: ```suggestion
        <button onClick={this.tenSecondToast}>Close after 10000ms</button>
```

Author: Poor editing, thank you

        <button onClick={this.normalToast}>Close after 5000ms (default)</button>
        <ToastContainer autoClose={5000} />
      </div>
    );
  }
}
```

For some of the common controls, the toast can be configured like so:

```js
class App extends Component {
  topRightWarning = () => toast('Warning!', {
    type: toast.TYPE.WARNING,
    position: toast.POSITION.TOP_RIGHT,

>     onClose: ({ props }) => window.alert('I warned you') // props: all props passed to rendered toast component

Reviewer 2: Maybe just omit the props argument since it's not essential to the example? i.e. `() => window.alert(...)`

Author: I think I can find a better way to include them
```suggestion
    onClose: ({ props }) => window.alert(`You just closed the ${props.type} message`) // props: all props passed to rendered toast component
```

  });

  render() {
    return (
      <button onClick={this.topRightWarning}>Warn me</button>
      <ToastContainer autoClose={5000} />
    );
  }
}
```

`react-toastify` supports built-in themes for toasts, such as `success`, `info`, and `warning`, among others.
It also provides intuitive positioning, that can be set from the toast or configured on the `ToastContainer`.
Additionally, callbacks like `onClose` and `onOpen` allow the toast to have intelligent effects when the user interacts with it.

Another extensible behavior `react-toastify` supports is updating a toast:

```js
class Update extends Component {
  toastId = null;

  notify = () => this.toastId = toast("Hello", { autoClose: false });

  update = () => toast.update(this.toastId, { type: toast.TYPE.INFO, autoClose: 5000 });

  render(){
      return (
        <div>
          <button onClick={this.notify}>Notify</button>
          <button onClick={this.update}>Update</button>
        </div>
      )
  }
}
```

As in this example, the `toastId` gives the developer control over the toast beyond configuration and display.

### Notable Features & Details

#### Toast ID

> the `toast()` function returns an ID of the toast it creates, which can be used to act upon that toast.

Reviewer 2: ```suggestion
The `toast()` function returns an ID of the toast it creates, which can be used to act upon that toast.
```

> For instance, with the toast ID the toast can be used with `dismiss()`, `update()`, and `isActive()`.

Reviewer 2: I wonder what the motivation is, for using IDs instead of JavaScript objects with methods.

Author: I'm not sure, I'm going to add a note to the takeaways to make sure that's noted
